sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/RocksDBStateStoreProvider.scala

```diff
@@ -0,0 +1,328 @@
+  @volatile private var hadoopConf: Configuration = _
+
+  private[sql] lazy val rocksDB = {
+    val dfsRootDir = stateStoreId.storeCheckpointLocation().toString
+    val storeIdStr = s"StateStoreId(opId=${stateStoreId.operatorId}," +
+      s"partId=${stateStoreId.partitionId},name=${stateStoreId.storeName})"
+    val sparkConf = Option(SparkEnv.get).map(_.conf).getOrElse(new SparkConf)
+    val localRootDir = Utils.createTempDir(Utils.getLocalDir(sparkConf), storeIdStr)
+    new RocksDB(dfsRootDir, RocksDBConf(storeConf), localRootDir, hadoopConf, storeIdStr)
+  }
+
+  private lazy val encoder = new StateEncoder
+
+  private def verify(condition: => Boolean, msg: String): Unit = {
+    if (!condition) { throw new IllegalStateException(msg) }
+  }
+
+  /**
+   * Encodes/decodes UnsafeRows to versioned byte arrays.
+   * It uses the first byte of the generated byte array to store the version the describes how the
+   * row is encoded in the rest of the byte array. Currently, the default version is 0,
+   *
+   * VERSION 0:  [ VERSION (1 byte) | ROW (N bytes) ]
+   *    The bytes of a UnsafeRow is written unmodified to starting from offset 1
+   *    (offset 0 is the version byte of value 0). That is, if the unsafe row has N bytes,
+   *    then the generated array byte will be N+1 bytes.
+   */
+  class StateEncoder {
+    import RocksDBStateStoreProvider._
+
+    // Reusable objects
+    private val keyRow = new UnsafeRow(keySchema.size)
+    private val valueRow = new UnsafeRow(valueSchema.size)
+    private val rowTuple = new UnsafeRowPair()
+
+    /**
+     * Encode the UnsafeRow of N bytes as a N+1 byte array.
+     * @note This creates a new byte array and memcopies the UnsafeRow to the new array.
+     */
+    def encode(row: UnsafeRow): Array[Byte] = {
+      val bytesToEncode = row.getBytes
+      val encodedBytes = new Array[Byte](bytesToEncode.length + STATE_ENCODING_NUM_VERSION_BYTES)
+      Platform.putByte(encodedBytes, Platform.BYTE_ARRAY_OFFSET, STATE_ENCODING_VERSION)
+      // Platform.BYTE_ARRAY_OFFSET is the recommended way to memcopy b/w byte arrays. See Platform.
+      Platform.copyMemory(
+        bytesToEncode, Platform.BYTE_ARRAY_OFFSET,
+        encodedBytes, Platform.BYTE_ARRAY_OFFSET + STATE_ENCODING_NUM_VERSION_BYTES,
+        bytesToEncode.length)
+      encodedBytes
+    }
+
+    /**
+     * Decode byte array for a key to a UnsafeRow.
+     * @note The UnsafeRow returned is reused across calls, and the UnsafeRow just points to
+     *       the given byte array.
+     */
+    def decodeKey(keyBytes: Array[Byte]): UnsafeRow = {
+      if (keyBytes != null) {
+        // Platform.BYTE_ARRAY_OFFSET is the recommended way refer to the 1st offset. See Platform.
+        keyRow.pointTo(
+          keyBytes,
+          Platform.BYTE_ARRAY_OFFSET + STATE_ENCODING_NUM_VERSION_BYTES,
+          keyBytes.length - STATE_ENCODING_NUM_VERSION_BYTES)
+        keyRow
+      } else null
```

Comment:
nit: I rarely see style like this in Spark.

Reply:
Make sense, done in the next commit.